cache/watcher.go

```diff
@@ -0,0 +1,46 @@
+package cache
+
+import (
+	"sync/atomic"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// watcher holds one client’s buffered stream of events.
+type watcher struct {
+	eventQueue chan *clientv3.Event
+	predicate  func([]byte) bool
+	stopped    int32
+}
+
+func newWatcher(bufSize int, pred KeyPredicate) *watcher {
+	return &watcher{
+		eventQueue: make(chan *clientv3.Event, bufSize),
+		predicate:  pred,
+	}
+}
+
+// true  -> event delivered or filtered/duplicate
+// false -> buffer full (caller should mark watcher “lagging”)
+func (w *watcher) enqueueEvent(event *clientv3.Event, nextRev *int64) bool {
```

Comment:
Don't return via arguments by using pointers. This is not something we do in Go.
Just accept nextRev argument and return newNextRev.